packages/endo/README.md

```diff
@@ -0,0 +1,283 @@
+  This is fitting because Endo runs Node _within_ a safe sandbox.
+  This is fitting in turn because Endo is built on the legacy of Google Caja.
+  In Spanish, "caja" means "box" and is related to the Latin word "capsum" and
+  English "capsule", as in "encapsulate".
+* Endo is an anagram of Node and Deno.
+  That is to say, we are not Done yet.
+* The `endo` command, like the `sudo` command, is a "do" command.
+  However, instead of escalating priviliedge, it encapsulates priviledge.
+* So, "E.N.Do" forms the acronym "Encapsulated Node Do".
+
+So, just as "soo-doo" (super user do) and "soo-doh" (like "pseudo") are valid
+pronunciations of `sudo`, "en-doh" and "en-doo" are both valid pronunciations of
+`endo`.
+
+
+## Package Descriptors
+
+Endo uses [Compartments] from the [SES] shim, one for each Node.js package
+your application needs.
+Endo generates a compartment graph from Node.js packaged module descriptors:
+the `package.json` files of the application and all its dependencies.
+Consequently, an application must have a `package.json`.
+
+Each package has its own descriptor, `package.json`.
+Some standard properties of the descriptor are relevant and used by a
+compartment map.
+
+* `name`
+* `type`
+* `main`
+* `exports`
+* `browser`
+* `dependencies`
+* `files`
+
+The compartment map will contain one compartment for each `package.json`
+necessary to build the application.
+Like Node.js, Endo trusts the package manager to arrange the packages such that
+a satisfactory version of every package's dependencies rests in a parent
+directory, under `node_modules`.
+
+The `main`, `browser`, and `exports` properties determine the modules each
+package exports to other compartments.
+
+The `exports` property describes [package entry points] and can be influenced
+by build _tags_.
+Currently, the only tag supported by Endo is `import`, which indicates that the
+module map should use ESM modules over CommonJS modules or other variants.
+
+> TODO
+>
+> A future version may reveal other tags like `browser` to prepare an
+> application for use in a web client.
+> For this case, Endo would prepare a JSON manifest like an `importmap` (if not
+> precisely an `importmap`).
+> The "compartment map" would be consistent except when the dependency graph
+> changes so updates could be automated with a `postinstall` script.
+> Preparing a web application for production would follow a process similar to
+> creating an archive, but with the `browser` build tag.
+
+The `browser` and `require` tags are well-known but not yet supported.
+The `browser` tag will apply for compartment maps generated for use on the web.
+The `require` tag is a fallback for environments that do not support ESM and
+will never apply.
+
+If no `exports` apply to the root of the compartment namespace (`"."`),
+the `main` property serves as a default.
+
+> TODO
+>
+> The absence of `exports` implies that _all_ of the modules in the package are
```

Comment:
I think this matches what we talked about, but I'll mention it here for the benefit of others.. In general, I think it's important that packages be able to restrict their callers to well-specified entrypoints. Once packages have authorities that their external callers do not, such packages will want to partition themselves internally. If a caller could pick-and-choose which of my modules to invoke, their authority is the maximum of all of my own module's exports, even if most of those modules are supposed to be reserved for my own modules to use. That would limit package author's flexibility to arrange their own code.

I think it's fine to have "no `exports`" means "export everything", since nobody in the past was using this sort of POLA partitioning. As long as eventually (perhaps immediately), packages which *do* provide an `exports` can be confident that code outside that package cannot invoke the package anywhere except those named entrypoints.

Reply:
Endo will enforce `exports` if present, for sure.